# long-read-RNA-seq-plotting-suite

**plotGeneCounts**
This function creates plots for the top 6 most upregulated and the top 6 most downregulated genes, using data generated by DESeq2. The plots show the number of times each gene is present in each sample, separated by group.

**plotDEvolcano**
This function creates two volcano plots of the differentially expressed genes outputted by DESeq2. Each plot has a different set of cut-off values.

**plotReadQualityHist**
This function creates a histogram of read basecall qualities before filtering out low basecall quality reads. The x-axis scale is logarithmic.

**plotReadLengthVsQualScatterhistLogData**
This function creates a scatterplot with marginal density histograms of the read length vs the basecall quality. The x-axis scale is logarithmic.

**plotReadLengthVsQualDensityLogData**
This function creates a density plot of the read length vs the basecall quality before filtering out low basecall quality reads. The x-axis scale is logarithmic.

**plotReadLengthHistLogData**
This function creates a log-transformed histogram of read lengths before filtering out low basecall quality reads. The data are log-transformed.

**plotReadLengthHistLogAxis**
This function creates a histogram of read lengths before filtering out low basecall quality reads. The x-axis scale is logarithmic.

**plotPercentIdentity**
This function creates a plot of Percent identity histograms.

**plotfeatureCountsSummaryGalaxy**
This function creates a plot of the results of featureCounts, when run on Galaxy.

**plotMAPQmapped**
This function creates a plot of Mapping quality (MAPQ) vs Number of mapped reads.

**plotMapVsUnmapLength**
This function creates a plot showing the distribution of the lengths of the mapped and unmapped reads separately.

**plotDEUvolcano**
This function creates a volcano plot of the differentially used exons outputted by DEXSeq.

**plotVariantCounts**
This function creates a plot of the splicing variant type counts for each group.

**plotGeneCountsGalaxy**
This function creates plots for the top 6 most upregulated and the top 6 most downregulated genes, using data generated by DESeq2. The plots show the number of times each gene is present in each sample, separated by group.

**plotNumberOfReads**
This function creates a plot of the Number of sequenced and Aligned reads.

**plotCompPercentAlignReads**
This function creates a plot comparing the percentage of reads that aligned at least once by two different aligners.

**plotCompNumAlignments**
This function creates a plot comparing the number of alignments generated by two different aligners.

**plotfeatureCountsSummary**
This function creates a plot of the results of featureCounts.

**plotSequencedReadLengthVsAlignedReadLength**
This function creates a plot of Sequenced read length vs Aligned read length.

**plotNumberOfMappings**
This function creates a plot of the Number of alignments for the mapped reads.

**plotMappingInfoPercent**
This function creates a plot of Percentage of mapped reads vs Type of alignment.

**plotMappingInfo**
This function creates a plot of the Number of mapped reads vs the Type of alignment.

**plotMapVsUnmapQual**
This function creates a plot showing the distribution of the basecall qualities of the mapped and unmapped reads separately.

**plotCompareAlignments**
This function creates a plot comparing the number of reads that aligned at least once by two different aligners.
